File: apps/server/src/services/rechargeService.ts
import prisma from "../config/prisma";
import { createError } from "../middleware/errorHandler";
import { logger } from "../utils/logger";

export interface RechargeInput {
  walletId: string;
  walletAlias: string;
  tokenSymbol: string;
  network: string;
  accountAddress: string;
  amount: string;
  memo?: string;
}

export interface RechargeDeviceInfo {
  deviceId: string;
  platform: string;
  version?: string;
}

export interface RechargeResult {
  id: string;
  walletId: string;
  walletAlias: string;
  accountAddress: string;
  tokenSymbol: string;
  tokenName: string;
  amount: string;
  memo: string;
  deviceId: string;
  platform: string;
  version: string;
  createdAt: Date;
}

export interface RechargeListFilter {
  page?: number;
  limit?: number;
  walletId?: string;
  tokenSymbol?: string;
}

/**
 * 充值：对系统内钱包的指定代币增加余额，并记录充值日志。
 * 仅允许 recharge_allowed_devices 配置中的设备操作。
 * 余额操作通过 wallet_subscriptions → wallets_addresses → assets_addresses (address_id) 链路。
 */
export async function recharge(
  input: RechargeInput,
  deviceInfo: RechargeDeviceInfo
): Promise<RechargeResult> {
  const { walletId, walletAlias, tokenSymbol, network, accountAddress, amount, memo } = input;

  logger.info("RECHARGE", `充值请求: walletId=${walletId}, tokenSymbol=${tokenSymbol}, amount=${amount}, deviceId=${deviceInfo.deviceId.slice(0, 8)}...`);

  // 1. 校验设备是否有充值权限
  const allowedConfig = await prisma.appConfig.findUnique({
    where: { key: "recharge_allowed_devices" },
  });

  let allowedDevices: string[] = [];
  if (allowedConfig?.value) {
    try {
      const parsed = JSON.parse(allowedConfig.value);
      if (Array.isArray(parsed)) {
        allowedDevices = parsed.filter((v) => typeof v === "string");
      }
    } catch {
      // JSON 解析失败，视为空列表
    }
  }

  if (!allowedDevices.includes(deviceInfo.deviceId)) {
    logger.warn("RECHARGE", `充值失败: 设备无权限 - deviceId=${deviceInfo.deviceId.slice(0, 8)}...`);
    throw createError(403, "当前设备无充值权限", "RECHARGE_DEVICE_NOT_ALLOWED");
  }

  // 2. 查找钱包
  const wallet = await prisma.wallet.findUnique({
    where: { id: walletId },
  });
  if (!wallet) {
    throw createError(404, "钱包不存在", "WALLET_NOT_FOUND");
  }

  // 3. 查找资产（symbol + chain 复合查询）
  const asset = await prisma.asset.findFirst({
    where: { symbol: tokenSymbol, chain: network },
  });
  if (!asset) {
    throw createError(404, "资产不存在", "ASSET_NOT_FOUND");
  }

  // 4. 通过 accountAddress 查找 wallets_addresses 记录
  const walletAddress = await prisma.walletAddress.findFirst({
    where: { address: accountAddress, chain: network },
  });
  if (!walletAddress) {
    throw createError(404, "该地址在系统中不存在", "ADDRESS_NOT_FOUND");
  }

  // 4-5. 余额更新 + 充值记录写入事务，防止并发充值导致余额不一致
  const [assetsAddress, record] = await prisma.$transaction(async (tx: any) => {
    // 4. 余额更新（upsert 在事务内）
    const aa = await tx.assetsAddress.upsert({
      where: {
        addressId_assetId: { addressId: walletAddress.id, assetId: asset.id },
      },
      update: {
        balance: { increment: parseFloat(amount) },
      },
      create: {
        addressId: walletAddress.id,
        assetId: asset.id,
        chain: network,
        balance: parseFloat(amount),
      },
    });

    logger.info("RECHARGE", `余额更新成功: assetsAddressId=${aa.id}, newBalance=${aa.balance}`);

    // 5. 写入充值记录
    const rec = await tx.recharge.create({
      data: {
        walletId,
        walletAlias: walletAlias || wallet.alias || wallet.id,
        accountAddress: accountAddress,
        tokenSymbol,
        tokenName: asset.name,
        amount: parseFloat(amount),
        memo: memo || "",
        deviceId: deviceInfo.deviceId,
        platform: deviceInfo.platform,
        version: deviceInfo.version || "",
      },
    });

    logger.info("RECHARGE", `充值成功: recordId=${rec.id}, wallet=${wallet.id}, token=${tokenSymbol}, amount=${amount}`);

    return [aa, rec];
  });

  return {
    id: record.id,
    walletId: record.walletId,
    walletAlias: record.walletAlias,
    accountAddress: record.accountAddress,
    tokenSymbol: record.tokenSymbol,
    tokenName: record.tokenName,
    amount: record.amount.toString(),
    memo: record.memo,
    deviceId: record.deviceId,
    platform: record.platform,
    version: record.version,
    createdAt: record.createdAt,
  };
}

/**
 * 分页查询充值记录
 */
export async function getRecharges(
  filter: RechargeListFilter
): Promise<{ recharges: RechargeResult[]; total: number }> {
  const page = filter.page || 1;
  const limit = Math.min(filter.limit || 20, 100);

  const where: any = {};
  if (filter.walletId) {
    where.walletId = filter.walletId;
  }
  if (filter.tokenSymbol) {
    where.tokenSymbol = filter.tokenSymbol;
  }

  const [records, total] = await Promise.all([
    prisma.recharge.findMany({
      where,
      orderBy: { createdAt: "desc" },
      skip: (page - 1) * limit,
      take: limit,
    }),
    prisma.recharge.count({ where }),
  ]);

  return {
    recharges: records.map((r: any) => ({
      id: r.id,
      walletId: r.walletId,
      walletAlias: r.walletAlias,
      accountAddress: r.accountAddress,
      tokenSymbol: r.tokenSymbol,
      tokenName: r.tokenName,
      amount: r.amount.toString(),
      memo: r.memo,
      deviceId: r.deviceId,
      platform: r.platform,
      version: r.version,
      createdAt: r.createdAt,
    })),
    total,
  };
}

File: apps/server/src/services/recharge_service.rs
//! 充值服务 — 迁移自 IMWallet services/rechargeService.ts

use crate::chain::address_validator;
use crate::db::query::{tx_exec, tx_query, tx_query_count, vals};
use crate::errors::AppError;
use crate::models::AppConfigEntity;
use rbatis::RBatis;
use rust_decimal::Decimal;
use serde::{Deserialize, Serialize};
use std::sync::Arc;

use crate::utils::short_addr;

#[derive(Debug, Deserialize)]
pub struct RechargeInput {
    pub wallet_id: String,
    pub wallet_alias: String,
    pub token_symbol: String,
    pub network: String,
    pub account_address: String,
    pub amount: Decimal,
    pub memo: Option<String>,
}
#[derive(Debug, Serialize)]
pub struct RechargeResult {
    pub id: String,
    pub wallet_id: String,
    pub wallet_alias: String,
    pub account_address: String,
    pub token_symbol: String,
    pub amount: Decimal,
}

pub async fn execute_recharge(
    rb: Arc<RBatis>,
    input: RechargeInput,
    device_id: &str,
    platform: &str,
    version: &str,
) -> Result<RechargeResult, AppError> {
    let t0 = std::time::Instant::now();
    // 校验充值设备白名单：仅白名单中的设备可充值，白名单为空时拒绝所有设备
    let t1 = std::time::Instant::now();
    let allowed: Vec<String> = crate::db::query::query_one::<AppConfigEntity>(
        &rb,
        "SELECT * FROM app_configs WHERE key = $1",
        vals!["recharge_allowed_devices"],
    )
    .await?
    .and_then(|c| serde_json::from_str::<Vec<String>>(&c.value).ok())
    .unwrap_or_default();
    log::debug!(
        "[耗时] recharge 查询白名单 {:.2}ms",
        t1.elapsed().as_millis() as f64
    );
    if allowed.is_empty() || !allowed.iter().any(|d| d == device_id) {
        log::warn!(
            "[充值] 拒绝 — 设备{}不在充值白名单中(白名单为空={})",
            short_addr(device_id),
            allowed.is_empty()
        );
        return Err(AppError::Forbidden("无权限进行充值操作".into()));
    }

    // 校验充值地址格式与链类型匹配
    let v = address_validator::validate_address_for_chain(&input.account_address, &input.network);
    if !v.is_valid {
        return Err(AppError::BadRequest(
            v.error.unwrap_or_else(|| "充值地址格式无效".into()),
        ));
    }

    // 使用事务保护整个充值操作
    let t2 = std::time::Instant::now();
    let tx = rb.acquire_begin().await?;
    log::debug!(
        "[耗时] recharge acquire_tx {:.2}ms",
        t2.elapsed().as_millis() as f64
    );

    // 从缓存获取资产元数据（启动时已预热），按 symbol + chain 查找
    let asset_map = crate::services::asset_service::get_cached_assets_map();
    let asset = asset_map
        .values()
        .find(|a| a.symbol == input.token_symbol && a.chain == input.network)
        .cloned()
        .ok_or_else(|| AppError::NotFound("代币类型不存在".into()))?;

    #[derive(serde::Deserialize)]
    struct A {
        id: String,
    }
    let addrs: Vec<A> = tx_query(&tx, "SELECT wa.id FROM wallet_subscriptions ws JOIN wallets_addresses wa ON wa.id = ws.address_id WHERE ws.wallet_id = $1 AND wa.chain = $2 AND wa.address = $3 LIMIT 1", vals![&input.wallet_id, &input.network, &input.account_address]).await?;
    let addr = addrs
        .into_iter()
        .next()
        .ok_or_else(|| AppError::NotFound("未找到该链上地址".into()))?;

    let cnt = tx_query_count(
        &tx,
        "SELECT COUNT(*) as cnt FROM assets_addresses WHERE address_id = $1 AND asset_id = $2",
        vals![&addr.id, &asset.id],
    )
    .await?;
    if cnt > 0 {
        tx_exec(&tx, "UPDATE assets_addresses SET balance = balance + $1, updated_at = NOW() WHERE address_id = $2 AND asset_id = $3", vals![rbdc::Decimal::new(&input.amount.to_string()).unwrap(), &addr.id, &asset.id]).await?;
    } else {
        tx_exec(&tx, "INSERT INTO assets_addresses (id, address_id, asset_id, chain, balance, created_at, updated_at) VALUES ($1, $2, $3, $4, $5, NOW(), NOW())", vals![uuid::Uuid::new_v4().to_string(), &addr.id, &asset.id, &input.network, rbdc::Decimal::new(&input.amount.to_string()).unwrap()]).await?;
    }

    let rid = uuid::Uuid::new_v4().to_string();
    tx_exec(&tx, "INSERT INTO recharges (id, wallet_id, wallet_alias, account_address, token_symbol, token_name, amount, memo, device_id, platform, version, created_at) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, NOW())", vals![&rid, &input.wallet_id, &input.wallet_alias, &input.account_address, &input.token_symbol, &asset.name, rbdc::Decimal::new(&input.amount.to_string()).unwrap(), input.memo.as_deref().unwrap_or(""), device_id, platform, version]).await?;

    tx.commit().await?;
    log::debug!(
        "[耗时] recharge 事务写入+commit {:.2}ms",
        t2.elapsed().as_millis() as f64
    );

    log::info!(
        "[充值] 完成 — 充值ID={}, 钱包{}(ID{}), 代币{}({}), 充值金额 {}, 设备{}, 总耗时 {:.2}ms",
        &rid,
        &input.wallet_alias,
        &input.wallet_id,
        &input.token_symbol,
        &input.network,
        input.amount,
        short_addr(device_id),
        t0.elapsed().as_millis() as f64
    );

    Ok(RechargeResult {
        id: rid,
        wallet_id: input.wallet_id,
        wallet_alias: input.wallet_alias,
        account_address: input.account_address,
        token_symbol: input.token_symbol,
        amount: input.amount,
    })
}
